eth/protocols/snap/sync.go

```diff
@@ -0,0 +1,718 @@
+		s.storageSynced += uint64(len(res.slots))
+		s.storageProofs += uint64(len(res.bounds))
+
+		// Hack
+		if res.nodes == nil {
+			break
+		}
+		// Persist every finalized trie node that's not on the boundary
+		it := res.nodes.NewIterator(nil, nil)
+		for it.Next() {
+			// Boundary nodes are not written, since they are incomplete
+			if _, ok := res.bounds[common.BytesToHash(it.Key())]; ok {
+				continue
+			}
+			// Node not a boundary, persist to disk
+			batch.Put(it.Key(), it.Value())
+			s.bloom.Add(it.Key())
+
+			s.storageNodes++
+			s.storageBytes += common.StorageSize(common.HashLength + len(it.Value()))
+		}
+		it.Release()
+
+		if err := batch.Write(); err != nil {
+			return false, err
+		}
+		batch.Reset()
+
+		// Storage range processed, step to the next chunk
+		log.Debug("Persisted range of storage slots", "account", account, "slot", res.last)
+		s.report(false)
+
+		// If the response contained all the data in one shot (no proofs), there
+		// is no reason to continue the sync, report immediate success.
+		if len(res.bounds) == 0 {
+			return true, nil
+		}
+		s.nextSlot = common.BigToHash(new(big.Int).Add(res.last.Big(), big.NewInt(1)))
+	}
+	return false, nil
+}
+
+// OnAccounts is a callback method to invoke when a range of accounts are
+// received from a remote peer.
+func (s *Syncer) OnAccounts(peer *Peer, id uint64, hashes []common.Hash, accounts [][]byte, proof [][]byte) error {
+	peer.Log().Trace("Delivering range of accounts", "hashes", len(hashes), "accounts", len(accounts), "proofs", len(proof))
+
+	// If the request is stale, discard it
+	s.lock.Lock()
+	req, ok := s.accountReqs[peer.id]
+	if !ok || req.id != id {
+		peer.Log().Warn("Unexpected account range packet")
+		s.lock.Unlock()
+		return nil
+	}
+	delete(s.accountReqs, peer.id)
+	root := s.root
+	s.lock.Unlock()
+
+	// If the response is unavailable snapshot, forward to the requester
+	if len(hashes) == 0 && len(accounts) == 0 && len(proof) == 0 {
+		select {
+		//case <-req.cancel:
+		case s.accountResps <- nil:
+		}
+		return nil
+	}
+	// Reconstruct a partial trie from the response and verify it
+	keys := make([][]byte, len(hashes))
+	for i, key := range hashes {
+		keys[i] = common.CopyBytes(key[:])
+	}
+	nodes := make(light.NodeList, len(proof))
+	for i, node := range proof {
+		nodes[i] = node
+	}
+	proofdb := nodes.NodeSet()
+
+	db, tr, err := trie.VerifyRangeProof(root, req.origin[:], keys, accounts, proofdb, proofdb)
+	if err != nil {
+		return err
+	}
+	// Partial trie reconstructed, send it to the scheduler for storage filling
+	bounds := make(map[common.Hash]struct{})
+
+	hasher := sha3.NewLegacyKeccak256()
+	for _, node := range proof {
+		hasher.Reset()
+		hasher.Write(node)
+		bounds[common.BytesToHash(hasher.Sum(nil))] = struct{}{}
+	}
+	response := &accountResponse{
+		hashes:   hashes,
+		accounts: accounts,
+		nodes:    db,
+		trie:     tr,
+		bounds:   bounds,
+	}
+	select {
+	//case <-req.cancel:
+	case s.accountResps <- response:
```

Comment:
hmm.. do we just drop this on the floor if the other side isn't listening? why would this happen? Shouldn't we at minimum log that we're dropping a batch of verified, useful data ? Or is it buffered?

Reply:
There's no `default` branch, so it blocks until it's read. The cancel case is also needed, but I need to implement some retry mechanism first, that's why I commented it out currently.